### PR TITLE
IDEMPIERE-5275 - Tabular Report Re-Query button/close paramater window

### PR DESCRIPTION
**Pull request by PeterTakacs300**
JIRA ticket: https://idempiere.atlassian.net/browse/IDEMPIERE-5275

### CODE REVIEW

**Comment by norbertbede** [comment]
notes:  Design, Architect: @igorpojzl  Coding/Commit: @PeterTakacs300

**Comment by norbertbede** [comment]
image

an example. tested many ways for community compactibility (the screencast was done on private version, the code is for latest v10)

**Review by hengsin — CHANGES_REQUESTED** [review]
2 issue with the rerun dialog
- the initial size of the dialog is not optimal, some options need horizontal scroll to access.
- the cancel button is not working correctly.

otherwise, seems to work just fine.

**Comment by norbertbede** [comment]
changes

1. sysconfig: ZK_REPORT_TABLE_OPEN_IN_NEW_TAB - default No - still possible to run re-run from toolbar - but doesn't confuse users with breaking change
2. cancel button fixed
3. scroll resolved - increased with. 

norbert

**Review by CarlosRuiz-globalqss — CHANGES_REQUESTED** [review]
Hi @PeterTakacs300 

Can you please include the following patch:
[1306_PeerReview.txt](https://github.com/idempiere/idempiere/files/8677136/1306_PeerReview.txt)


It adds the following changes:
- change the SysConfig scope to Tenant
- change the text of the message to "Run it again with new parameters"
- add the definition of the new button in Toolbar (this way it can be restricted per role)
- change timestamp in lang-addon.xml (because a css was changed)
- reorder font-icons.css
- reorder MSysConfig and generate serialVersionUID
- remove a warning on ReportCtl

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/ZkReportViewer.java [hunk 1]
```diff
@@ -1217,6 +1243,14 @@ else if (event.getName().equals(ON_RENDER_REPORT_EVENT))
 	        	this.onCtrlKeyEvent(keyEvent);
         	}
 		}
+        else if (event.getTarget() instanceof ProcessModalDialog)
+        {
+        	if (!DialogEvents.ON_WINDOW_CLOSE.equals(event.getName())){
+        		return;
+        	}
+        	
+        	hideBusyMask();
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/ZkReportViewer.java:1252]: if this is means to run hideBusyMask() if event is on_window_close, it is more readable to use a positive condition block instead, i.e if (DialogEvents.ON_WINDOW_CLOSE.equals(event.getName())) hideBusyMask();

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/ZkReportViewer.java [hunk 2]
```diff
@@ -1458,6 +1494,28 @@ private void cmd_report()
 		postRenderReportEvent();
 	}	//	cmd_report
 
+	/**
+	 * Refresh With Parameters
+	 */
+	private void cmd_reRun() {
+		int AD_Process_ID = m_reportEngine.getPrintInfo() != null ? m_reportEngine.getPrintInfo().getAD_Process_ID() : 0;
+		ProcessInfo pi = new ProcessInfo("RefreshWithParameters", AD_Process_ID);
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/ZkReportViewer.java:1502]: can AD_Process_ID be 0 here ? if yes, then the method should stop immediately (i.e return;) if AD_Process_ID is 0

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/ZkReportViewer.java [hunk 3]
```diff
@@ -1458,6 +1494,28 @@ private void cmd_report()
...
 
+	/**
+	 * Refresh With Parameters
+	 */
+	private void cmd_reRun() {
+		int AD_Process_ID = m_reportEngine.getPrintInfo() != null ? m_reportEngine.getPrintInfo().getAD_Process_ID() : 0;
+		ProcessInfo pi = new ProcessInfo("RefreshWithParameters", AD_Process_ID);
+		pi.setIsReportOverride();
+		ProcessModalDialog processModalDialog = new ProcessModalDialog(this, m_WindowNo, pi, false, true);
+		processModalDialog.setWidth("850px");
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/window/ZkReportViewer.java:1505]: use ZkUpdateUtil.setWindowWidthX(...) instead.

File: org.adempiere.base/src/org/compiere/process/ProcessInfo.java [hunk 4]
```diff
@@ -151,7 +152,10 @@ public ProcessInfo (String Title, int AD_Process_ID)
 	
 	/**	Export File				*/
 	private File				m_exportFile = null;
-
+	
+	/** Report Override Tab */
+	private boolean				m_IsReportOverride = false;
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/process/ProcessInfo.java:157]: I think it is more readable to use m_IsReplaceTabContent instead

File: org.adempiere.base/src/org/compiere/process/ProcessInfo.java [hunk 5]
```diff
@@ -938,5 +942,13 @@ public void setProcessUI(IProcessUI processUI) {
...
 		return processUI;
 	}
+
+	public boolean isReportOverride() {
+		return m_IsReportOverride;
+	}
+
+	public void setIsReportOverride() {
+		this.m_IsReportOverride = !(MSysConfig.getBooleanValue(MSysConfig.ZK_REPORT_TABLE_OPEN_IN_NEW_TAB, false, Env.getAD_Client_ID(Env.getCtx())));
+	}
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/process/ProcessInfo.java:952]: please always add javadoc for public methods. also, as per my comments above, I think it is more readable to use isReplaceTabContent and setReplaceTabContent instead.

File: org.adempiere.base/src/org/compiere/print/ReportEngine.java [hunk 6]
```diff
@@ -2390,6 +2392,14 @@ public String getReportType()
...
 	}
 	
+	public boolean isOverride() {
+		return m_isOverride;
+	}
+
+	public void setIsOverride(boolean m_isOverride) {
+		this.m_isOverride = m_isOverride;
+	}
+
```
Comment (hengsin) [inline org.adempiere.base/src/org/compiere/print/ReportEngine.java:2402]: add javadoc for public methods and I think it is more readable to use IsReplaceTabContent instead of isOverride.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/apps/ProcessModalDialog.java [hunk 7]
```diff
@@ -68,6 +69,28 @@ public ProcessModalDialog(int WindowNo, ProcessInfo pi, boolean autoStart)
 		this(null, WindowNo, pi, autoStart);
 	}
 	
+	/**
+	 * @param WindowNo
+	 * @param pi
+	 * @param autoStart
+	 */
+	public ProcessModalDialog(EventListener<Event> listener, int WindowNo, ProcessInfo pi, boolean autoStart, boolean loadSavedParams)
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/apps/ProcessModalDialog.java:77]: it seems this is always call with autoStart = false and loadSavedParams = true, if that's the case, simpler and more readable to just remove the 2 process parameter and use the false and true directly.

**Review by hengsin — CHANGES_REQUESTED** [review]

**Comment by PeterTakacs300** [comment]
I think I've applied all the requested changes, thank you for the patch and the detailed review @CarlosRuiz-globalqss and @hengsin

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/apps/ProcessModalDialog.java [hunk 8]
```diff
@@ -68,6 +69,26 @@ public ProcessModalDialog(int WindowNo, ProcessInfo pi, boolean autoStart)
 		this(null, WindowNo, pi, autoStart);
 	}
 	
+	/**
+	 * @param WindowNo
+	 * @param pi
+	 * @param autoStart
+	 */
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/apps/ProcessModalDialog.java:76]: wrong javadoc

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/Window.java [hunk 9]
```diff
@@ -59,6 +59,8 @@ public class Window extends org.zkoss.zul.Window implements ISupportMask
     public static final String INSERT_END = "insertEnd";
     /*** Insert next to the active tab of the tabbed window container ***/
     public static final String INSERT_NEXT = "insertNext";
+    /*** Insert override existing content ***/
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/component/Window.java:62]: Comment should be changed to something like "Replace Current Tab Content"

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/part/WindowContainer.java [hunk 10]
```diff
@@ -635,6 +639,38 @@ public Tab insertAfter(Tab refTab, Component comp, String title, boolean closeab
...
+     * @param refTab 
+     * @param comp
+     * @param title
+     * @return
+     */
+    public Tab replace(Tab refTab, Window comp, String title) {
+    	 
+         if (refTab == null)  
+         {
+         	tabbox.getTabs().getChildren();
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/part/WindowContainer.java:653]: what is this for ? Since refTab must not be null here, probably should just throw IllegalArgumentException here.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/part/WindowContainer.java [hunk 11]
```diff
@@ -635,6 +639,38 @@ public Tab insertAfter(Tab refTab, Component comp, String title, boolean closeab
...
+         	org.zkoss.zul.Tabpanel refpanel = refTab.getLinkedPanel();
+         	Component oldWindow = refpanel.getFirstChild();
+     		if(oldWindow instanceof ProcessDialog)
+     			((ProcessDialog)oldWindow).unlockUI(null);
+     		else if(oldWindow instanceof ZkReportViewer)
+				((ZkReportViewer)oldWindow).hideBusyMask();
+			else if(oldWindow instanceof AbstractADWindowContent)
+				((AbstractADWindowContent)oldWindow).hideBusyMask();
+     		((Window) oldWindow).onClose();
+     		comp.setParent(refpanel);
```
Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/part/WindowContainer.java:666]: the change make for this block is wrong. The previous if (oldWindow instanceof Window) condition is correct, what I mean is you should have the else part too, i.e when oldWindow is not instanceof Window, you should just detach oldWindow (probably sounds more readable if that's rename to firstChild instead) and attach comp to refpanel.

**Review by hengsin — CHANGES_REQUESTED** [review]

**Comment by hengsin** [comment]
@PeterTakacs300 , please resolve the conflict.

**Comment by norbertbede** [comment]
@hengsin, @PeterTakacs300 is out of office will be after 14h CET.

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/apps/ProcessDialog.java [hunk 12]
```diff
@@ -312,10 +312,11 @@ private void hideBusyMask()
 	{
 		if (mask != null && mask.getParent() != null) {
 			mask.detach();
-			StringBuilder script = new StringBuilder("(function(){let w=zk.Widget.$('#");
-			script.append(getParent().getUuid()).append("');w.busy=false;");
-			script.append("})()");
-			Clients.response(new AuScript(script.toString()));
+			if(getParent() != null) {
+				StringBuilder script = new StringBuilder("var w=zk.Widget.$('#");
```
Comment (CarlosRuiz-globalqss) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/apps/ProcessDialog.java:316]: WARNING: this is reverting the change from:
https://github.com/idempiere/idempiere/commit/895923f7f229e3c3195c691afdba3c12e6fe61ca#diff-45852062765555bb780d1af8808a6582852fa6a3102af104344b02c5b2bad61c

Comment (hengsin) [inline org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/apps/ProcessDialog.java:316]: this merge with master is wrong (i.e it should be using the (function(){...})()).

**Review by hengsin — CHANGES_REQUESTED** [review]

File: org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/apps/ProcessDialog.java (hunk 12 above)
Comment (PeterTakacs300) [reply org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/apps/ProcessDialog.java:316]: yes thank you, fixing in a moment..

**Review by CarlosRuiz-globalqss — APPROVED** [review]
from my side it works fine

let's wait for @hengsin approval too

**Review by hengsin — APPROVED** [review]
looks good and tested by Carlos.
